### PR TITLE
Feature/k8s provider process raw spec

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [ ] ~Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?~
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

Implemented raw k8s yaml format of k8s spec processing component for CAAS broker to make `k8s-raw-spec` hook command working.

TODO items:

- resource spec scheme validation needs to be implemented;
- Scale, Constraints, FileSystems, Devices, InitContainer for actions, etc. are not supported due to lack of the above TODO 1;

## QA steps

In a K8s charm, run 

```python
    layer.caas_base.k8s_raw_set(k8s_raw_spec)
```

the `k8s_raw_spec` could be valid k8s spec (multi resources supported as well);

```yaml
# sample
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
  namespace: ns-will-be-overwritten  # this namespace will be changed by Juju to the current model's namespace
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 8000
          protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  name: mock
  labels:
    app: mock
spec:
  ports:
  - port: 99
    protocol: TCP
    targetPort: 9949
  selector:
    app: mock
```

```sh
## bootstrap with feature flag enabled
$ juju bootstrap microk8s k1 --config='features=[raw-k8s-spec]'

$ juju add-model t1 microk8s

$ juju deploy /tmp/charm-builds/mariadb-k8s/ --debug  --resource mysql_image=mariadb --storage database=100M

$ mkubectl -nt1 get svc,deploy -o json | jq '.items[] | {metadata: .metadata, kind:.kind} | {name: .metadata.name,labels: .metadata.labels, annotations: .metadata.annotations, kind: .kind} '
{
  "name": "mariadb-k8s-operator",
  "labels": {
    "juju-operator": "mariadb-k8s"
  },
  "annotations": {
    "juju-version": "2.8-beta1",
    "juju.io/controller": "4a339ac1-9513-4ba6-84eb-a5fc57478ddc",
    "juju.io/model": "0548fd5f-1838-4dd3-8c67-0cbf52abb931"
  },
  "kind": "Service"
}
{
  "name": "mock",
  "labels": {
    "app": "mock",
    "juju-app": "mariadb-k8s"
  },
  "annotations": {
    "juju.io/controller": "4a339ac1-9513-4ba6-84eb-a5fc57478ddc",
    "juju.io/model": "0548fd5f-1838-4dd3-8c67-0cbf52abb931"
  },
  "kind": "Service"
}
{
  "name": "nginx-deployment",
  "labels": {
    "app": "nginx",
    "juju-app": "mariadb-k8s"
  },
  "annotations": {
    "deployment.kubernetes.io/revision": "1",
    "juju.io/controller": "4a339ac1-9513-4ba6-84eb-a5fc57478ddc",
    "juju.io/model": "0548fd5f-1838-4dd3-8c67-0cbf52abb931"
  },
  "kind": "Deployment"
}

$ juju remove-application mariadb-k8s -m k1:t1 --destroy-storage --force

$ mkubectl -nt1 get svc,deploy -o json | jq '.items[] | {metadata: .metadata, kind:.kind} | {name: .metadata.name,labels: .metadata.labels, annotations: .metadata.annotations, kind: .kind} '

```

## Documentation changes

Yes

## Bug reference

None

### CODE REVIEW

**Comment by ycliuhw** [comment]
[k8s-raw-spec provider implementation](https://trello.com/c/LLYgggFy/1095-k8s-raw-spec-provider-implementation)

File: caas/kubernetes/provider/k8s.go [hunk 1]
```diff
@@ -877,15 +904,54 @@ func processConstraints(pod *core.PodSpec, appName string, cons constraints.Valu
...
+func (k *kubernetesClient) applyRawK8sSpec(
+	appName string,
+	statusCallback caas.StatusCallbackFunc,
+	params *caas.ServiceParams,
+	numUnits int,
+	config application.ConfigAttributes,
+) (err error) {
+
+	if params == nil || len(params.RawK8sSpec) == 0 {
+		return errors.Errorf("missing raw pod spec")
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go:918]: raw k8s spec

File: caas/kubernetes/provider/specs/builder.go [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+	}
+	if err := d.validateWorkload(); err != nil {
+		return errors.Trace(err)
+	}
+	// TODO(caas): check if service resource type matches the raw service spec.
+	// TODO(caas): get the API scheme and do validation further.
+	return nil
+}
+
+func (d *deployer) Deploy(ctx context.Context, spec string, force bool) error {
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/builder.go:218]: comment

File: caas/kubernetes/provider/specs/builder.go [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+		logger.Tracef("gvk.GroupKind() %s, gvk.Version %q, item.mapping %s", pretty.Sprint(gvk.GroupKind()), gvk.Version, pretty.Sprint(item.mapping))
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		d.resources = append(d.resources, item)
+	}
+}
+
+func (d deployer) apply(ctx context.Context, wg *sync.WaitGroup, info resourceInfo, force bool, errChan chan<- error) (err error) {
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/builder.go:342]: probably worth a small comment, same with load()

File: caas/kubernetes/provider/specs/builder_test.go [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+var rawK8sSpec = `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+  labels:
+    app: nginx
+  namespace: ns-will-be-overwritten
+spec:
+  replicas: 3
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/builder_test.go:65]: How do we handle a mismatch between this and num units the user asks for?
I think num units should override this

File: caas/kubernetes/provider/k8s.go [hunk 5]
```diff
@@ -877,15 +904,54 @@ func processConstraints(pod *core.PodSpec, appName string, cons constraints.Valu
...
 		return errors.Trace(err)
 	}
-	logger.Debugf("specStr -> %s", specStr)
-	logger.Debugf("spec -> %s", pretty.Sprint(spec))
-	return errors.NotImplementedf("raw k8s spec")
+
+	logger.Debugf("creating/updating application %s", appName)
+	deploymentName := k.deploymentName(appName, false)
+
+	if numUnits < 0 {
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go:936]: numUnits is checked here but not used below when applying the raw spec.
I think it should set the replicas count

Comment (ycliuhw) [reply caas/kubernetes/provider/k8s.go:936]: As we discussed and the TODO 2 mentioned, the `.Spec` schema is not parsed in this PR yet which causes a few key features that are missing: Scale, Constraints, FileSystems, Devices, InitContainer for actions etc.

File: worker/caasunitprovisioner/deployment_worker.go [hunk 6]
```diff
@@ -193,22 +182,12 @@ func (w *deploymentWorker) loop() error {
...
-			ResourceTags:      info.Tags,
-			Filesystems:       info.Filesystems,
-			Devices:           info.Devices,
-			OperatorImagePath: info.OperatorImagePath,
-			Deployment: caas.DeploymentParams{
-				DeploymentType: caas.DeploymentType(info.DeploymentInfo.DeploymentType),
-				ServiceType:    caas.ServiceType(info.DeploymentInfo.ServiceType),
-			},
+		serviceParams, err := provisionInfoToServiceParams(info)
+		logger.Warningf("provisionInfoToServiceParams info -> len(info.PodSpec) %d, len(info.RawK8sSpec) %d", len(info.PodSpec), len(info.RawK8sSpec))
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply worker/caasunitprovisioner/deployment_worker.go:187]: ops, removed thanks

File: worker/caasunitprovisioner/deployment_worker.go [hunk 7]
```diff
@@ -193,22 +182,12 @@ func (w *deploymentWorker) loop() error {
...
+		if err != nil {
+			logger.Warningf("provisionInfoToServiceParams info -> %s", pretty.Sprint(info))
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply worker/caasunitprovisioner/deployment_worker.go:189]: yeah, removed because there is an error returned.

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**